### PR TITLE
Collect webview messages async

### PR DESCRIPTION
**Pull request by mattseddon**
Follow up to https://github.com/iterative/vscode-dvc/pull/4402#discussion_r1281384894

This PR makes the collection of experiments and plots webview messages async.

### CODE REVIEW

File: extension/src/test/suite/experiments/index.test.ts [hunk 1]
```diff
@@ -1527,85 +1522,49 @@ suite('Experiments Test Suite', () => {
...
-      const messageSpy = spy(BaseWebview.prototype, 'show')
-
-      const resourceLocator = disposable.track(
-        new ResourceLocator(extensionUri)
-      )
-      stub(dvcReader, 'stageList').resolves('train')
-      stub(dvcReader, 'dag').resolves('')
-
-      const pipeline = buildExperimentsPipeline({
+      const { experiments, messageSpy } = await buildExperimentsWebview({
```
Comment (mattseddon) [inline extension/src/test/suite/experiments/index.test.ts:1525]: [F] After reading this test I had no idea why it didn't use the builder so moved it across.

File: extension/src/test/suite/experiments/index.test.ts [hunk 2]
```diff
@@ -997,21 +989,22 @@ suite('Experiments Test Suite', () => {
...
-        })
+        }
+      )
 
       messageSpy.resetHistory()
 
       const movedGroup = 'params:params.yaml'
 
-      const tableChangePromise = experimentsUpdatedEvent(experiments)
+      const messageSent = waitForSpyCall(messageSpy, messageSpy.callCount)
```
Comment (mattseddon) [inline extension/src/test/suite/experiments/index.test.ts:1002]: [F] I had to add this new util as the data for the message is now gathered async.

Comment (mattseddon) [reply extension/src/test/suite/experiments/index.test.ts:1002]: Where possible we still rely on the old util (experimentsUpdatedEvent(experiments)`)

File: extension/src/plots/webview/messages.ts [hunk 3]
```diff
@@ -61,18 +61,12 @@ export class WebviewMessages {
...
-      cliError: this.errors.getCliError()?.error || null,
-      comparison: this.getComparisonPlots(),
-      custom: this.getCustomPlots(),
-      hasPlots: !!this.paths.hasPaths(),
-      hasUnselectedPlots: this.paths.getHasUnselectedPlots(),
-      sectionCollapsed: this.plots.getSectionCollapsed(),
-      selectedRevisions,
-      template: this.getTemplatePlots(selectedRevisions)
-    })
+    const webview = this.getWebview()
```
Comment (mattseddon) [inline extension/src/plots/webview/messages.ts:64]: [F] although this function is duplicated it is not worth the hassle that a base class would bring

**Review by julieg18 — APPROVED** [review]
Great work!

**Pull request merged**